src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/LaunchProfiles/AddLaunchProfileAction.cs

```diff
@@ -0,0 +1,34 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.Debug;
+using Microsoft.VisualStudio.ProjectSystem.Query;
+using Microsoft.VisualStudio.ProjectSystem.Query.ProjectModelMethods.Actions;
+using Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Query
+{
+    /// <summary>
+    /// <see cref="IQueryActionExecutor"/> handling <see cref="ProjectModelActionNames.AddLaunchProfile"/> actions.
+    /// </summary>
+    internal sealed class AddLaunchProfileAction : LaunchProfileActionBase
+    {
+        private readonly AddLaunchProfile _executableStep;
+
+        public AddLaunchProfileAction(AddLaunchProfile executableStep)
+        {
+            _executableStep = executableStep;
+        }
+
+        protected override async Task ExecuteAsync(ILaunchSettingsProvider launchSettingsProvider)
+        {
+            await launchSettingsProvider.AddOrUpdateProfileAsync(
```

Comment:
Could just return the task to avoid the state machine. `RemoveLaunchProfileAction` does this.